File: 15-3sum/15-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # if nums length is 0, then return []
        if len(nums) == 0 or len(nums) == 1:
            return []
        # if nums length is 1, return [] 
        
        # brute force way => use three nested loops, find all possible combinations of sum(nums[i and j and k]) == target
        
        nums.sort()
        # if all the same number, then no need to go through the solution, because no same number other than 0 can add to 0
        answer = []
        for i in range(0, len(nums)-1):
            # check for dup
            if i > 0 and nums[i] == nums[i-1]:
                continue
        
            l = i + 1
            r = len(nums)-1
            target = 0 - nums[i]

            
            while r > l:
                if nums[r] + nums[l] > target:
                    r-=1 
                elif nums[r] + nums[l] < target:
                    l+=1
                else:
                    answer.append([nums[i], nums[l], nums[r]])
                    r-=1 
                    l+=1
                    
                    while (r > l and nums[l] == nums[l-1]):
                        l+=1 
                    while (r > l and nums[r] == nums[r+1]):
                        r-=1
            
        return answer
                        
                    
                    
                
                
        
        
        
        

File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    // no length no iterate
    if (nums.length  === 0) {
        return []
    }
    nums = nums.sort((a,b)=>a-b)

    const triplets = []
    
    // iterate nums
    // if current num is same as prev num
        // continue
    
    for (let i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i] === nums[i-1]) {
            continue
        }
        
        let l = i + 1
        let r = nums.length-1
        
        while (l < r) {
            sum = nums[l] + nums[r] + nums[i]
            if (sum > 0) {
                // make it smaller
                r --
            } else if (sum < 0) {
                // make it bigger
                l ++
            } else {
                // found answer
                triplets.push([nums[i], nums[l], nums[r]])
                l ++ 
                // make sure left num is not same as prev left num
                while (l < r && nums[l] === nums[l-1]) {
                    l++
                }
            }
        }
    }
    
    return triplets
};